src/front/glsl/lex_pp.rs

```diff
@@ -0,0 +1,253 @@
+use super::{parser::Token, token::TokenMetadata, types::parse_type};
+use crate::FastHashMap;
+use pp_rs::{
+    pp::Preprocessor,
+    token::{Punct, Token as PPToken, TokenValue},
+};
+use std::collections::VecDeque;
+
+//#[derive(Clone, Debug)]
+pub struct LexerPP<'a> {
+    pp: Preprocessor<'a>,
+    pp_buf: VecDeque<PPToken>,
+}
+
+impl<'a> LexerPP<'a> {
+    pub fn new(input: &'a str, defines: &'a FastHashMap<String, String>) -> Self {
+        let mut pp = Preprocessor::new(input);
+        for (define, value) in defines {
+            pp.add_define(define, value).unwrap(); //TODO: handle error
+        }
+        LexerPP {
+            pp,
+            pp_buf: Default::default(),
+        }
+    }
+
+    pub fn next(&mut self) -> Option<Token> {
+        let mut meta = TokenMetadata {
+            line: 0,
+            chars: 0..0,
+        };
+        let buf_token = self.pp_buf.pop_front();
+        let pp_token = if let Some(t) = buf_token {
+            Some(t)
+        } else {
+            let t = self.pp.next();
+            if let Some(t) = t {
+                match t {
+                    Ok(t) => Some(t),
+                    Err((err, loc)) => {
+                        meta.line = loc.line as usize;
+                        meta.chars.start = loc.pos as usize;
+                        //TODO: proper location end
+                        meta.chars.end = loc.pos as usize + 1;
+                        return Some(Token::Unknown((meta, format!("{:?}", err))));
+                    }
+                }
```

Comment:
You could write this as
```rust
let pp_token = self.pp_buf.pop_front().or_else(|| {
   Some(match self.pp.next()? {
     Ok(t) => t,
     Err((err, loc)) => {
         meta.line = loc.line as usize;
         meta.chars.start = loc.pos as usize;
         //TODO: proper location end
         meta.chars.end = loc.pos as usize + 1;
         Token::Unknown((meta, format!("{:?}", err)))
     }
   })
});
```
You could also add a `?` at the end and you just need to wrap the result in `Some` instead of doing a `map`

Reply:
nice, was struggling a bit with this part - had actually tried using `or_else` but of course `return` inside there wouldn't work.

Reply:
hmmm actually this won't work, as `Token::Unknown((meta, format!("{:?}", err)))` is a parser token, not a pp token :S

Reply:
you're right completely missed the return still something like this is in my opinion better
```rust
    let pp_token = match self.pp_buf.pop_front() {
        Some(t) => t,
        None => match self.pp.next()? {
            Ok(t) => t,
            Err((err, loc)) => {
                meta.line = loc.line as usize;
                meta.chars.start = loc.pos as usize;
                //TODO: proper location end
                meta.chars.end = loc.pos as usize + 1;
                return Some(Token::Unknown((meta, format!("{:?}", err))));
            }
        }
    };
```
you wouldn't need the `map` with the closure this way and is more compact and readable in my opinion